resolver.go

```diff
@@ -128,3 +137,176 @@ func (a *serviceAddr) Network() string {
 func (a *serviceAddr) String() string {
 	return net.JoinHostPort(a.addr, strconv.Itoa(a.port))
 }
+
+// LookupServiceFunc is the signature of functions that can be used to lookup
+// service names.
+type LookupServiceFunc func(context.Context, string) ([]net.Addr, error)
+
+// The ResolverCache type provides the implementation of a caching layer for
+// service name resolutions.
+//
+// Instances of ResolverCache are save to use concurrently from multiple
+// goroutines.
+type ResolverCache struct {
+	// The maximum age of cache entries. If zero, a default value of 1 second is
+	// used.
+	Timeout time.Duration
+
+	once sync.Once
+	cmap *resolverCacheMap
+}
+
+// LookupService resolves a service name by fetching the address list from the
+// cache, or calling lookup if the name did not exist.
+func (cache *ResolverCache) LookupService(ctx context.Context, name string, lookup LookupServiceFunc) ([]net.Addr, error) {
+	cache.once.Do(cache.init)
+
+	now := time.Now()
+	timeout := cache.timeout()
+	entry, locked := cache.cmap.lookup(name, now, now.Add(timeout))
+
+	if locked {
+		// TODO: check the error type here and discard things like context
+		// cancellations and timeouts?
+		entry.addrs, entry.err = lookup(ctx, name)
```

Comment:
We could, however I think it would lead to more lock contentions. The purpose of having a mutex on each entry to to allow multiple service names to be looked up in parallel, while synchronizing all goroutines that are trying to lookup the same service name to avoid bursts of requests when an entry expires (or on the first lookup).
If we use a single lock then we lose concurrency on consul requests, or we may submit bursts of requests when an entry expires.